hw/ip/prim/dv/prim_prince/crypto_dpi_prince/prince_ref.h

```diff
@@ -0,0 +1,342 @@
+/*
+Copyright 2016 Sebastien Riou
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*! \file prince_ref.h
+    \brief Reference implementation of the Prince block cipher, complient to
+   C99. 'Reference' here means straightforward, unoptimized, and checked against
+   the few test vectors provided in the original paper
+   (http://eprint.iacr.org/2012/529.pdf). By defining the macro PRINCE_PRINT,
+   one can print out all successive internal states.
+*/
+
+/*
+ * This C-implementation of the PRINCE block cipher is taken from Sebastien
+ * Riou's open-sourced 'prince-c-ref' GitHub repository, found here:
+ * https://github.com/sebastien-riou/prince-c-ref/blob/master/include/prince_ref.h.
+ *
+ * Several modifications of varying severity have been made to this C model to
+ * enable extra external parameterizations for more thorough verification of the
+ * OpenTitan PRINCE hardware implementation.
+ * These modifications, in no particular order, are:
+ *
+ *    - Conversion of `prince_enc_dec_uint64_t(...)` to a non-static function to
+ *      allow usage by an external DPI-C model.
+ *    - Additional arguments added to `prince_enc_dec_uint64_t(...)`,
+ *      `prince_enc_dec(...)`, `prince_encrypt(...)` `prince_decrypt(...)`, and
+ *      `prince_core(...)` to allow an external DPI-C model to control the
+ *      number of cipher half-rounds and to toggle between the "legacy" key
+ *      schedule detailed in the original PRINCE paper and a newer key schedule.
+ *    - Modification of `prince_core(...)` to handle the new key schedule and
+ *      user-specified number of half-rounds.
+ */
+
+#ifndef OPENTITAN_HW_IP_PRIM_DV_PRIM_PRINCE_CRYPTO_DPI_PRINCE_PRINCE_REF_H_
+#define OPENTITAN_HW_IP_PRIM_DV_PRIM_PRINCE_CRYPTO_DPI_PRINCE_PRINCE_REF_H_
+
+#include <stdint.h>
+#include <string.h>
+
+#ifndef PRINCE_PRINT
+#define PRINCE_PRINT(a) \
+  do {                  \
+  } while (0)
+#endif
+
+/**
+ * Converts a byte array into a 64 bit integer
+ * byte at index 0 is placed as the most significant byte.
```

Comment:
(Optional) I understand that this is third-party code, but if you have a free minute (:smile:), would you mind following a style that helps doxygen to extract the API docs better?

```
Summary line

Long(er) description
```

Reply:
Sure, I'll re-format the existing comments - thanks for the suggestion!